### PR TITLE
[3.11] gh-103921: Improve typing documentation (GH-104642)

### PR DESCRIPTION
**Pull request by JelleZijlstra**
(cherry picked from commit 060277d96bf4ba86df8e4d65831a8cbdfeb51fc5)

This backports some of the changes from the PEP 695 documentation PR (#104642) to 3.11, because I ended up making some improvements that are also applicable to older versions of typing:

- Consistent indentation of examples in `typing.rst`
- Document the call signature of TypeVar, ParamSpec, TypeVarTuple
- Omit mention of the variance-related parameters of ParamSpec, which are not meaningful in the type system
- Some wording improvements

* Issue: gh-103921

----
:books: Documentation preview :books:: https://cpython-previews--105007.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -423,8 +430,8 @@ to this is that a list of types can be used to substitute a :class:`ParamSpec`::
    >>> Z[int, [dict, float]]
    __main__.Z[int, (<class 'dict'>, <class 'float'>)]
 
-
-Furthermore, a generic with only one parameter specification variable will accept
+Another difference between :class:`TypeVar` and :class:`ParamSpec` is that a
+generic with only one parameter specification variable will accept
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: Not sure this needs to be changed in 3.11. The reason why we changed it in 3.12+ was because we added another paragraph immediately preceding this one, which broke the flow; starting the paragraph with "Furthermore" didn't work anymore. But in 3.11, we're not adding the new paragraph, so "Furthermore" works fine, and is more concise. I'd revert this change:

```suggestion
Furthermore, a generic with only one parameter specification variable will accept
```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -1275,177 +1283,203 @@ These are not used in annotations. They are building blocks for creating generic
...
 
-    Using a *constrained* type variable, however, means that the ``TypeVar``
-    can only ever be solved as being exactly one of the constraints given::
+   .. attribute:: __name__
+
+      The name of the type variable.
+
+   .. attribute:: __covariant__
+
+      Whether the type var has been explicitly marked as covariant.
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: Since there's no such thing as implicit {co,contra,in}variance on 3.11:

```suggestion
      Whether the type var has been marked as covariant.
```

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -1275,177 +1283,203 @@ These are not used in annotations. They are building blocks for creating generic
...
+
+   .. attribute:: __contravariant__
 
-       a = concatenate('one', 'two')
-       reveal_type(a)  # revealed type is str
+      Whether the type var has been explicitly marked as contravariant.
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      Whether the type var has been marked as contravariant.
```

File: Doc/library/typing.rst [hunk 4]
```diff
@@ -1275,177 +1283,203 @@ These are not used in annotations. They are building blocks for creating generic
...
 
-    Type variable tuples must *always* be unpacked. This helps distinguish type
-    variable tuples from normal type variables::
+   Type variable tuples can be used in the same contexts as normal type
+   variables. For example, in class definitions, arguments, and return types::
 
-        x: Ts          # Not valid
-        x: tuple[Ts]   # Not valid
-        x: tuple[*Ts]  # The correct way to do it
+      class Array(Generic[*Shape]):
```
Comment (AlexWaygood) [inline Doc/library/typing.rst:1434]: What is this, 3.12? :p

The `Shape` TypeVarTuple is now undefined here, with your changes:

```suggestion
      Shape = TypeVarTuple("Shape")
      class Array(Generic[*Shape]):
```

File: Doc/library/typing.rst [hunk 5]
```diff
@@ -1275,177 +1283,203 @@ These are not used in annotations. They are building blocks for creating generic
...
 
-        class Array2(Generic[*Shape, DType]):  # This would also be fine
-            pass
+   However, note that at most one type variable tuple may appear in a single
+   list of type arguments or type parameters::
 
-        float_array_1d: Array[float, Height] = Array()     # Totally fine
-        int_array_2d: Array[int, Height, Width] = Array()  # Yup, fine too
+      x: tuple[*Ts, *Ts]            # Not valid
+      class Array(Generic[*Shape, *Shape]):  # Not valid
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: These comments are aligned in the current version of the 3.11 docs, and it looks more attractive that way

```suggestion
      x: tuple[*Ts, *Ts]                     # Not valid
      class Array(Generic[*Shape, *Shape]):  # Not valid
```

**Review by AlexWaygood — APPROVED** [review]

File: Doc/library/typing.rst [hunk 6]
```diff
@@ -1275,177 +1282,204 @@ These are not used in annotations. They are building blocks for creating generic
...
 
-        Shape = TypeVarTuple('Shape')
-        class Array(Generic[*Shape]):
-            def __getitem__(self, key: tuple[*Shape]) -> float: ...
-            def __abs__(self) -> "Array[*Shape]": ...
-            def get_shape(self) -> tuple[*Shape]: ...
+      DType = TypeVar('DType')
 
-    Type variable tuples can be happily combined with normal type variables::
+      class Array[DType, *Shape]:  # This is fine
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: Some PEP-695 syntax has slipped in here, and in another place

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: Doc/library/typing.rst (hunk 6 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst]:
```suggestion
      class Array(Generic[DType, *Shape]):  # This is fine
```

File: Doc/library/typing.rst [hunk 7]
```diff
@@ -1275,177 +1282,204 @@ These are not used in annotations. They are building blocks for creating generic
...
+          pass
 
-        DType = TypeVar('DType')
+      class Array2[*Shape, DType]:  # This would also be fine
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      class Array2(Generic[*Shape, DType]):  # This would also be fine
```

File: Doc/library/typing.rst (hunk 6 above)
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: I only see this one and the one right below it, are there others?

Comment (AlexWaygood) [reply Doc/library/typing.rst]: Yeah I think that's all 👍

Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Thanks, fixed both of those

**Review by AlexWaygood — APPROVED** [review]
Thanks!
